simulator/src/output_settings.rs

```diff
@@ -0,0 +1,112 @@
+use crate::{display::SimulatorDisplay, theme::BinaryColorTheme};
+use embedded_graphics::{
+    geometry::{Point, Size},
+    pixelcolor::PixelColor,
+    DrawTarget,
+};
+
+/// Output settings.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct OutputSettings {
+    /// Pixel scale.
+    pub scale: u32,
+    /// Spacing between pixels.
+    pub pixel_spacing: u32,
+    /// Binary color theme.
+    pub theme: BinaryColorTheme,
+}
+
+impl OutputSettings {
+    /// Calculates the size of the framebuffer required to display the scaled display.
+    pub(crate) fn framebuffer_size<C>(&self, display: &SimulatorDisplay<C>) -> Size
+    where
+        C: PixelColor,
+    {
+        let width = display.size().width;
+        let height = display.size().height;
+        let output_width = width * self.scale + (width - 1) * self.pixel_spacing;
+        let output_height = height * self.scale + (height - 1) * self.pixel_spacing;
+
+        Size::new(output_width, output_height)
+    }
+
+    /// Translates a output coordinate to the corresponding display coordinate.
+    pub(crate) const fn output_to_display(&self, output_point: Point) -> Point {
+        let pitch = self.pixel_pitch() as i32;
+        Point::new(output_point.x / pitch, output_point.y / pitch)
+    }
+
+    pub(crate) const fn pixel_pitch(&self) -> u32 {
+        self.scale + self.pixel_spacing
+    }
+}
+
+impl Default for OutputSettings {
+    fn default() -> Self {
+        Self {
+            scale: 1,
+            pixel_spacing: 0,
+            theme: BinaryColorTheme::Default,
+        }
+    }
+}
+
+/// Output settings builder.
+pub struct OutputSettingsBuilder {
+    scale: Option<u32>,
+    pixel_spacing: Option<u32>,
+    theme: BinaryColorTheme,
+}
+
+impl OutputSettingsBuilder {
+    /// Creates new output settings builder.
+    pub fn new() -> Self {
+        Self {
+            scale: None,
+            pixel_spacing: None,
+            theme: BinaryColorTheme::Default,
+        }
+    }
+
+    /// Sets the pixel scale.
+    ///
+    /// A scale of `2` or higher is useful for viewing the simulator on high DPI displays.
+    pub fn scale(mut self, scale: u32) -> Self {
+        if scale == 0 {
+            panic!("scale must be >= 0");
+        }
```

Comment:
[`NonZeroU32`](https://doc.rust-lang.org/beta/std/num/struct.NonZeroU32.html)?

Reply:
I did think about this, but having to use `.scale(NonZeroU32::new(2).unwrap())` to call the method seems a bit cumbersome IMO.

Reply:
Yeah that API change is a bit ugly. You could convert it inside the method though. Not bothered either way though. Feel free to resolve this thread if you cba to change it :)

Reply:
After trying to use `NonZeroU32` I decided to not use it at the moment and instead just document that the function panics if the scale is set to `0`.